src/greynirseq/nicenlp/tasks/translation_with_glossary.py

```diff
@@ -0,0 +1,407 @@
+#
+# Based on fairseq/tasks/translation.py that has the following license
+#
+#    Copyright (c) Facebook, Inc. and its affiliates.
+#
+#    This source code is licensed under the MIT license found in the
+#    LICENSE file in the root directory of this source tree.
+
+import argparse
+import logging
+import random
+from dataclasses import dataclass
+from functools import partial
+from typing import Callable, Dict, List, Optional
+
+import torch
+from fairseq.data import Dictionary, encoders
+from fairseq.data.base_wrapper_dataset import BaseWrapperDataset
+from fairseq.data.language_pair_dataset import LanguagePairDataset
+from fairseq.tasks import register_task
+from torch.functional import Tensor
+
+from greynirseq.nicenlp.data.encoding import get_word_beginnings
+from greynirseq.nicenlp.tasks.translation_with_backtranslation import TranslationWithBacktranslationTask
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class GlossaryTaskConfig:
+    """ Configuration for the glossary task."""
+
+    enabled: bool
+    ok_to_increase_dict_size: bool
+    constraint_positional_shift: int
+    seq_sample_ratio: float
+    mean_whole_word: float
+    stddev_whole_word: float
+
+    @staticmethod
+    def from_args(args):
+        # It is not ok to increase the dictionary size if we are loading a pretrained model.
+        # TODO: verify that this is the case.
+        ok_to_increase_dict_size = args.restore_file is None
+        return GlossaryTaskConfig(
+            enabled=args.glossary_enabled,
+            ok_to_increase_dict_size=ok_to_increase_dict_size,
+            constraint_positional_shift=args.glossary_constraint_positional_shift,
+            seq_sample_ratio=args.glossary_seq_sample_ratio,
+            mean_whole_word=args.glossary_mean_whole_word,
+            stddev_whole_word=args.glossary_stddev_whole_word,
+        )
+
+    @staticmethod
+    def add_args(parser: argparse.ArgumentParser) -> None:
+        """Add arguments to the parser for this task."""
+        parser.add_argument(
+            "--glossary-enabled", default=False, action="store_true", help="Should the glossary task be enabled?"
+        )
+        parser.add_argument(
+            "--glossary-constraint-positional-shift",
+            default=1024,
+            type=int,
+            help="By how much should we shift the positional encoding of the glossary constraints?",
```

Comment:
Let's test starting from 0 (+ pad_idx since fairseq does that automatically). Then we do not need to think too carefully about learnt positional encodings (if it works).